lib/src/main/kotlin/com/swmansion/starknet/provider/gateway/GatewayProvider.kt

```diff
@@ -122,10 +122,9 @@ class GatewayProvider(
 
     private fun transformResponseForTransactionAndSerialize(body: String): Transaction {
         try {
-            val transformedResponse = json.parseToJsonElement(body).jsonObject["transaction"]
-            val encodedResponse = json.encodeToString(transformedResponse)
-            return json.decodeFromString(TransactionPolymorphicSerializer, encodedResponse)
-        } catch (e: IllegalArgumentException) {
+            val transformedResponse = json.parseToJsonElement(body).jsonObject.getValue("transaction")
```

Comment:
have you tried combining transfirming and polymorphic serializers? I guess this would be an ideal way to refactor. I have attempted it while doind rpc2.1 but failed, and did want to spend too much time on it;

With your proposed changes we still convert string to json, and then again to string.

Reply:
I actually the same issues a while before rpc2.1. But I can try it again.